docs/engineering.md

```diff
@@ -0,0 +1,595 @@
+| `user-disable`   | `su.endStudy("user-disable")`   | Implies user uninstalled or disabled add-on, or (BUG) Normandy uninstalled it.              |
+| `expired`        | `su.endStudy("expired")`        | Time-limited study reached expiration.                                                     |
+| `ended-positive` | `su.endStudy("ended-positive")` | General study-defined 'good ending', such as attempting to use feature.                    |
+| `ended-negative` | `su.endStudy("ended-negative")` | General study-defined 'bad ending', such as clicking 'I do not like this feature'.         |
+| `ended-neutral`  | `su.endStudy("ended-neutral")`  | General study-defined 'neutral ending'.                                                    |
+| `ineligible`     | `su.endStudy("ineligible")`     | During install, client actually not appropriate for study, for some study-specific reason. |
+| **EXIT**         |                                 |
+| `exit`           |                                 | automatically sent as part of `endStudy`                                                   |
+
+**Note**: Every user should have
+
+* exactly 1 each of ENTER, EXIT
+* exactly 1 of either INSTALL or INELGIBLE
+* exactly one 'ending' ping (which might be INELIGIBLE, EXPIRED, USER-DISABLE, ENDED-\*)
+
+**Note**: [Full Schemas - gregglind/shield-study-schemas](https://github.com/gregglind/shield-study-schemas/tree/master/schemas-client)
+
+
+### Send your own probes
+
+Use: `shieldStudy.telemetry(anObjectWithStringValues)`
+
+This will send data to the `shield-study-addon` bucket. The `key=>string` map will be the `payload.data.attributes` key.
+
+Example:
+
+```javascript
+// values must be strings
+studyUtils.telemetry({ evt: "click", button: "share", times: "3" });
+```
+
+### Defining Custom Study Endings
+
+Suppose you want some 'early endings', such as:
+
+* positive: user reached "end of the built UI".
+* negative: user clicked on "no thanks".
+
+Define in `endings`:
+
+```
+endings: {
+  /** User defined endings */
+  "user-attempted-signup": {
+    "baseUrl": "http://www.example.com/?reason=too-popular",
+    "study_state": "ended-positive",  // neutral is default
+  }
+}
+```
+
+Then:
+
+```
+studyUtils.endStudy("user-attempted-signup");
+```
+
+## Viewing Sent Telemetry Probes
+
+### client
+
+1. **Use the QA Helper Add-on**
+
+   The QA-Shield-Study-Helper lists the `payload.data` field for every `shield-study` and `shield-study-addon` ping.
+
+   [Bugzilla for QA Helper Add-on](https://bugzilla.mozilla.org/show_bug.cgi?id=1407757)
+   [direct install link for Signed XPI for @qa-shield-study-helper-1.0.0.xpi][qa-helper-addon-direct]
+
+   Example output:
+
+   ```text
+   // common fields
+
+   branch        up-to-expectations-1        // should describe Question text
+   study_name    57-perception-shield-study
+   addon_version 1.0.0
+   version       3
+   ```
+
+
```

Comment:
We're missing an opening code fence here. The rendered Markdown payloads all look sketchy.